### PR TITLE
Adding Ingredient entity

### PR DESCRIPTION
**Pull request by DianaMD21**
This branch contains the Ingredient entity. An ingredient has a product, the measure unit and the amount of that measure unit that will be used in a specific plate. For example, a burguer will contain the following basic ingredients: 2 slices of bread, 1 pound of meat, 100 grams of lettuce (...). Please check the diagram.

### CODE REVIEW

File: src/entities/Ingredient.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package entities;
+
+import enums.MeasureUnit;
+
+public class Ingredient extends BaseEntity<Long>{
+    private Product product;
+    private MeasureUnit measureUnit;
+    private Long amount;
```
Comment (reynoldmorel) [inline src/entities/Ingredient.java:8]: I think `amount` can be double as well

**Review by reynoldmorel — CHANGES_REQUESTED** [review]
In general I think you need to relate products and ingredients?

**Comment by DianaMD21** [comment]
> In general I think you need to relate products and ingredients?

FinalProduct will be a class that extends from Product, and final product will have a list of ingredients

**Comment by reynoldmorel** [comment]
> > In general I think you need to relate products and ingredients?
> 
> FinalProduct will be a class that extends from Product, and final product will have a list of ingredients

Could you elaborate?

File: src/entities/Ingredient.java (hunk 1 above)
Comment (DianaMD21) [reply src/entities/Ingredient.java]: solved

**Comment by DianaMD21** [comment]
> > > In general I think you need to relate products and ingredients?
> > 
> > 
> > FinalProduct will be a class that extends from Product, and final product will have a list of ingredients
> 
> Could you elaborate?

A product can be soomething like a type of bread, soda, single items that do not require preparation, for example. Salt, Flour, Meat, those are products. But not all products are items that can be shown in the menu and selled. A FinalProduct (that extends from Product) will be a plate or meal, that contains a list of ingredients. For example, a FinalProduct can be spaguetti, that has a list of ingredidents of: 500 grams of pasta, 300 grams of meatballs, 100 grams of tomato sauce, and more. An ingredient especifies the product, measure unit and amount of that measure unit that will be used to prepare a plate

**Comment by reynoldmorel** [comment]
> > > > In general I think you need to relate products and ingredients?
> > > 
> > > 
> > > FinalProduct will be a class that extends from Product, and final product will have a list of ingredients
> > 
> > 
> > Could you elaborate?
> 
> A product can be soomething like a type of bread, soda, single items that do not require preparation, for example. Salt, Flour, Meat, those are products. But not all products are items that can be shown in the menu and selled. A FinalProduct (that extends from Product) will be a plate or meal, that contains a list of ingredients. For example, a FinalProduct can be spaguetti, that has a list of ingredidents of: 500 grams of pasta, 300 grams of meatballs, 100 grams of tomato sauce, and more. An ingredient especifies the product, measure unit and amount of that measure unit that will be used to prepare a plate

Why Pasta, meatballs and tomato sauce can't be products as well? Also I think bread and soda measurable items. Not sure the about difference of these 2 concepts of your application

**Comment by DianaMD21** [comment]
> > > > > In general I think you need to relate products and ingredients?
> > > > 
> > > > 
> > > > FinalProduct will be a class that extends from Product, and final product will have a list of ingredients
> > > 
> > > 
> > > Could you elaborate?
> > 
> > 
> > A product can be soomething like a type of bread, soda, single items that do not require preparation, for example. Salt, Flour, Meat, those are products. But not all products are items that can be shown in the menu and selled. A FinalProduct (that extends from Product) will be a plate or meal, that contains a list of ingredients. For example, a FinalProduct can be spaguetti, that has a list of ingredidents of: 500 grams of pasta, 300 grams of meatballs, 100 grams of tomato sauce, and more. An ingredient especifies the product, measure unit and amount of that measure unit that will be used to prepare a plate
> 
> Why Pasta, meatballs and tomato sauce can't be products as well? Also I think bread and soda measurable items. Not sure the about difference of these 2 concepts of your application

I was even thinking of removing the MeasureUnit attribute in Product, so that a product only has a name attrbitute. Because a product will be used in ingredient, and also in stock, and even on menu items. The measureUnit of a product in Ingredient can be different than the measureUnit used in stock, the amounts will also be different. A product is only an item, the name of it, even if is something as simple as a meatball, or as complex as a full pasta plate.

It's a way to separate them, so that i wont have issues in the future when working with menu items and stock. An ingredient will have a product, and the measure unit and amount that will be used of that product in an specific plate.

For example, a 1/2 Pound Burger will have an meat ingredient of that amount (1/2 a pound). But a 1 Pound Special Burguer will have 1 pound of meat as an ingredient: same products (meat), different amounts.

**Comment by reynoldmorel** [comment]
I think achieve the same thing with products.

One thing is what you have and another thing is what you need in order to prepare a final product.

What you have is on stock. What you need is on the order the client request.

When the order is applied then you decrease the stock availability for the product, no matter what product since your final product defines what it needs to be prepared.

A relationship between final product and product can tell you how much of a product is needed to be prepared and this relationship can point to single unique product.

I'm still not sure why ingredient is needed since product is something that you will also sell.

Soda is not a product is a final product, assuming final products are the products that actually sell to a client

**Comment by DianaMD21** [comment]
> I think you can achieve the same thing with products.
> 
> One thing is what you have and another thing is what you need in order to prepare a final product.
> 
> What you have is on stock. What you need is on the order the client request.
> 
> When the order is applied then you decrease the stock availability for the product, no matter what product since your final product defines what it needs to be prepared.
> 
> A relationship between final product and product can tell you how much of a product is needed to be prepared and this relationship can point to single unique product.
> 
> I'm still not sure why ingredient is needed since product is something that you will also sell.
> 
> Soda is not a product is a final product, assuming final products are the products that actually sell to a client

What i dont understand is how i will manage the amount of each product when creating a FinalProduct
If two different plates require the same product but with different amounts, how will it work? 2 instances of the same product will be created with different amounts?

How will i know what amount of each product to use when creating a plate?

**Comment by reynoldmorel** [comment]
Let me know if this is clear now:

image

**Pull request closed without merging**